<a href="https://colab.research.google.com/github/Techvolutions05/DS-Unit-1-Sprint-2-Statistics/blob/master/module4/Anell_Santos_LS_DS_124_Introduction_to_Bayesian_Inference_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lambda School Data Science Module 124

## Introduction to Bayesian Inference




## Assignment - Code it up!

We used pure math to apply Bayes Theorem to drug tests. Now write Python code to reproduce the results! This is purposefully open ended - you'll have to think about how you should represent probabilities and events. You can and should look things up.

Specific goals/targets:

### 1) Write a function 

`def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):` 

You should only truly need these three values in order to apply Bayes Theorem. In this example, imagine that individuals are taking a breathalyzer test with an 8% false positive rate, a 100% true positive rate, and that our prior belief about drunk driving in the population is 1/1000. 
 - What is the probability that a person is drunk after one positive breathalyzer test?
 - What is the probability that a person is drunk after two positive breathalyzer tests?
 - How many positive breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit?


In [0]:
# TODO - code!
import pandas as pd
from scipy import stats


In [4]:
def prob_drunk_given_positive(prob_drunk_prior, false_positive_rate, true_positive_rate):
 
  num = true_positive_rate * prob_drunk_prior
  mar =  (true_positive_rate * prob_drunk_prior) + (false_positive_rate * (1 - prob_drunk_prior))
  post = num / mar 
  return post

# prob_drunk_prior = 1/1000
# true positive rate = 1 (100%)
post = prob_drunk_given_positive(0.001, 0.08, 1)
post

0.012357884330202669

The probability that a person is drunk after one positive breathalyzer test is 1.2%

In [6]:
post2 = prob_drunk_given_positive(post, 0.08, 1)
print(post2)

0.13525210993291495


The probability that a person is drunk after one positive breathalyzer test is 13%

In [8]:
post3 = prob_drunk_given_positive(post2, 0.08, 1)
print(post3)

0.6615996951348605


The probability that a person is drunk after one positive breathalyzer test is 66%

In [9]:
post4 = prob_drunk_given_positive(post3, 0.08, 1)
print(post4)

0.9606895076105054


The probability that a person is drunk after one positive breathalyzer test is 96%

About 4 breathalyzer tests are needed in order to have a probability that's greater than 95% that a person is drunk beyond the legal limit.

---

## Idea 1:

### You
What do you care about?

What do you know about?

What decisions do you face?




### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:**

**Trend related to the news:**

**Who does that?:**

**Answering a question people care about:**

**Valuable to businesses:**

**What's the most popular?:**

**Cost/Money rankings:**

### Misconceptions

What misconceptions do people have about your topic?

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

###Data

Where could you search for data about your topic?

---

## Idea 2:

### You
What do you care about?

What do you know about?

What decisions do you face?



### Seven templates
Training Kit (https://learn.lambdaschool.com/ds/module/recedjanlbpqxic2r) explains the seven templates from Priceonomics.

Can you apply the templates to your topics?

**Geographic Variation:**

**Trend related to the news:**

**Who does that?:**

**Answering a question people care about:**

**Valuable to businesses:**

**What's the most popular?:**

**Cost/Money rankings:**

### Misconceptions

What misconceptions do people have about your topic?

### Examples

What data storytelling example inspires you?

Could you do a new hypothesis, for the same question?

Could you do a new question, for the same topic?

Could you do a new topic, with the same "style"?

###Data

Where could you search for data about your topic?

## Resources

- [Worked example of Bayes rule calculation](https://en.wikipedia.org/wiki/Bayes'_theorem#Examples) (helpful as it fully breaks out the denominator)
- [Source code for mvsdist in scipy](https://github.com/scipy/scipy/blob/90534919e139d2a81c24bf08341734ff41a3db12/scipy/stats/morestats.py#L139)

## Stretch Goals:

- Go back and study the content from Modules 1 & 2 & 3 to make sure that you're really comfortable with them.
- Apply a Bayesian technique to a problem you previously worked (in an assignment or project work) on from a frequentist (standard) perspective
- Check out [PyMC3](https://docs.pymc.io/) (note this goes beyond hypothesis tests into modeling) - read the guides and work through some examples
- Take PyMC3 further - see if you can build something with it!